### ISSUE TITLE
torch.compile overflows torch.sinh near float32 boundary while eager returns finite values

### ISSUE THREAD

**Issue by ALinrunrun** [pytorch/pytorch#183763]
### 🐛 Describe the bug

`torch.compile` with the Inductor backend produces different overflow behavior from eager execution for `torch.sinh` on large float32 inputs near the overflow boundary.

Eager execution returns finite float32 values around `1e38`, while the compiled version returns `inf` / `-inf` for the same inputs.

```
import warnings
warnings.filterwarnings("ignore")

import numpy as np
import torch

def fn(x):
    return torch.sinh(x)

x = torch.tensor([88.85, 89.2, 88.95, -89.0], dtype=torch.float32)

eager = fn(x).cpu().numpy()

torch._dynamo.reset()
compiled = torch.compile(fn, backend="inductor", fullgraph=True)(x).cpu().numpy()

print(f"input: {x.cpu().numpy()}")
print(f"eager: {eager}")
print(f"comp : {compiled}")

assert np.isfinite(eager[0]) and eager[0] > 1e38
assert np.isinf(comp[0])
```

### Error logs

input: [ 88.85  89.2   88.95 -89.  ]
eager: [ 1.9321198e+38  2.7418043e+38  2.1353193e+38 -2.2448064e+38]
comp : [ inf  inf  inf -inf]

### Versions

PyTorch version: 2.13.0.dev20260513+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.11.15 (main, Mar 11 2026, 17:20:07) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.17.0-20-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration:
GPU 0: NVIDIA RTX 6000 Ada Generation
GPU 1: NVIDIA RTX 6000 Ada Generation

Nvidia driver version: 570.211.01
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] torch==2.13.0.dev20260513+cpu
[conda] numpy                      2.4.4                   pypi_0           pypi
[conda] torch                      2.13.0.dev20260513+cpu  pypi_0           pypi

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by karthickai** [comment]
**🤖 PT2 Repro Agent**

I tried to reproduce this issue on PyTorch nightly (`2.13.0.dev20260518+cu130`) on NVIDIA H100. The behavior is flaky (1/3 runs failed).

Bug reproduced intermittently across runs with TORCH_INDUCTOR_DISABLE_CACHE=1. On runs that fail: eager torch.sinh on float32 input [88.85, 89.2, 88.95, -89.0] returns finite values [1.93e+38, 2.74e+38, 2.14e+38, -2.24e+38] while compiled output returns [inf, inf, inf, -inf]. On runs that pass: compiled output matches eager exactly (max_diff=0.0, all finite). Observed 8/10 fail, 2/10 pass in a 10-run sample; 1/3 fail in a separate fresh 3-run cycle.

<details>
<summary>Error output</summary>

```
torch version: 2.13.0.dev20260518+cu130
cuda available: True
input:    [ 88.85  89.2   88.95 -89.  ]
eager:    [ 1.9321198e+38  2.7418043e+38  2.1353193e+38 -2.2448064e+38]
compiled: [ inf  inf  inf -inf]
eager_finite:    [ True  True  True  True]
compiled_finite: [False False False False]
finiteness_mismatch: True
test_fn result: FAIL (bug reproduced)

ERROR conda.cli.main_run:execute(49): `conda run python /home/karthickps/.repro_agent/workdir/pytorch_pytorch/183763/repro.py` failed. (See above for error)

```
</details>

<details>
<summary>Repro script I used</summary>

```python
import os
import warnings
warnings.filterwarnings("ignore")

import numpy as np
import torch


def fn(x):
    return torch.sinh(x)


def test_fn() -> bool:
    """Return True if NO bug (eager matches compiled). False if bug present.

    Repro for pytorch/pytorch#183763: torch.compile (Inductor) returns
    inf/-inf for torch.sinh on float32 inputs near the overflow boundary,
    while eager returns finite values around 1e38.
    """
    torch._dynamo.reset()
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")

    x = torch.tensor([88.85, 89.2, 88.95, -89.0], dtype=torch.float32)

    eager_out = fn(x)
    compiled_out = torch.compile(fn, backend=backend, fullgraph=True)(x)

    eager_finite = torch.isfinite(eager_out)
    comp_finite = torch.isfinite(compiled_out)
    if not torch.equal(eager_finite, comp_finite):
        return False

    if eager_finite.all() and comp_finite.all():
        return bool(torch.allclose(eager_out, compiled_out, rtol=1e-3, atol=1e30))
    return True


if __name__ == "__main__":
    import traceback

    print(f"torch version: {torch.__version__}")
    print(f"cuda available: {torch.cuda.is_available()}")

    # Run test_fn FIRST so its compile is not contaminated by any earlier
    # compile in this process. (Doing a diagnostic compile before test_fn
    # changes inductor's behavior and intermittently hides the overflow.)
    try:
        ok = test_fn()
    except Exception:
        traceback.print_exc()
        os._exit(2)

    # Diagnostic: re-run eager + compiled in a fresh dynamo state to show the
    # actual output values to the reader.
    torch._dynamo.reset()
    x = torch.tensor([88.85, 89.2, 88.95, -89.0], dtype=torch.float32)
    eager = fn(x).cpu().numpy()
    compiled = torch.compile(fn, backend="inductor", fullgraph=True)(x).cpu().numpy()
    eager_finite = np.isfinite(eager)
    comp_finite = np.isfinite(compiled)
    print(f"input:    {x.cpu().numpy()}")
    print(f"eager:    {eager}")
    print(f"compiled: {compiled}")
    print(f"eager_finite:    {eager_finite}")
    print(f"compiled_finite: {comp_finite}")
    print(f"finiteness_mismatch: {bool(np.any(eager_finite != comp_finite))}")
    if eager_finite.all() and comp_finite.all():
        print(f"max_diff: {float(np.max(np.abs(eager - compiled)))}")

    print(f"test_fn result: {'PASS (no bug)' if ok else 'FAIL (bug reproduced)'}")
    os._exit(0 if ok else 1)

```
</details>

<details>
<summary>My environment</summary>

```
Collecting environment information...
PyTorch version: 2.13.0.dev20260518+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-14)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:13:31) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.16.1-0_fbk2_hardened_0_gf40efc324cc8-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA H100
Nvidia driver version: 580.126.09
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  46
On-line CPU(s) list:                     0-45
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      1
Core(s) per socket:                      46
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                4792.79
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean flushbyasid pausefilter pfthreshold v_vmsave_vmload vgif vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid fsrm flush_l1d arch_capabilities
Virtualization:                          AMD-V
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               2.9 MiB (46 instances)
L1i cache:                               2.9 MiB (46 instances)
L2 cache:                                23 MiB (46 instances)
L3 cache:                                736 MiB (46 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-45
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Vulnerable
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:                Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx==13.0.85
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] onnx==1.21.0
[pip3] onnx-ir==0.2.1
[pip3] onnxscript==0.7.0
[pip3] torch==2.13.0.dev20260517+cu130
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0.dev20260517+cu130
[pip3] torchdata==0.11.0
[pip3] torchinfo==1.8.0
[pip3] torchmetrics==1.9.0
[pip3] torchrec==1.7.0.dev20260413+cu130
[pip3] torchtune==0.6.1
[pip3] torchvision==0.28.0.dev20260518+cu130
[pip3] triton==3.7.0+git88b227e2
[conda] Could not collect
```
</details>

---
If the repro agent produced an incorrect result (wrong classification, bad repro script, missed the actual bug, etc.), add the label `repro:agent:h100:incorrect` to the issue.
